File: hgctl/src/client/executor.ts
import * as grpc from '@grpc/grpc-js';
import * as protoLoader from '@grpc/proto-loader';
import { Logger } from '../logger';
import { Context } from '../config/context';
import path from 'path';

// Load proto definitions from local copy
const PROTO_PATH = path.join(__dirname, '../protos/executor.proto');

const packageDefinition = protoLoader.loadSync(PROTO_PATH, {
  keepCase: true,
  longs: String,
  enums: String,
  defaults: true,
  oneofs: true
});

const proto = grpc.loadPackageDefinition(packageDefinition) as any;

export interface Performer {
  performer_id: string;
  avs_address: string;
  status: string;
  application_healthy: boolean;
  resource_healthy: boolean;
  artifact_digest?: string;
}

export interface DeployArtifactRequest {
  avsAddress: string;
  digest: string;
  registryUrl: string;
}

export interface DeployArtifactResponse {
  success: boolean;
  message?: string;
  performerId?: string;
}

export class ExecutorClient {
  private client: any;
  private logger: Logger;

  constructor(address: string, logger: Logger) {
    this.logger = logger;
    
    // Use insecure credentials for local development
    const credentials = address.includes('localhost') || address.includes('executor:')
      ? grpc.credentials.createInsecure()
      : grpc.credentials.createSsl();
    
    this.client = new proto.eigenlayer.hourglass.v1.ExecutorService(
      address,
      credentials
    );
  }

  async listPerformers(avsAddress?: string): Promise<Performer[]> {
    return new Promise((resolve, reject) => {
      const deadline = new Date();
      deadline.setSeconds(deadline.getSeconds() + 30);
      
      this.client.ListPerformers(
        { avs_address: avsAddress || '' },
        { deadline },
        (error: grpc.ServiceError | null, response: any) => {
          if (error) {
            reject(this.translateError(error, 'list performers'));
          } else {
            resolve(response.performers || []);
          }
        }
      );
    });
  }

  async deployArtifact(request: DeployArtifactRequest): Promise<DeployArtifactResponse> {
    this.logger.info(`Deploying artifact for AVS ${request.avsAddress}`);
    
    return new Promise((resolve, reject) => {
      const deadline = new Date();
      deadline.setSeconds(deadline.getSeconds() + 300); // 5 minutes
      
      this.client.DeployArtifact(
        {
          avs_address: request.avsAddress,
          digest: request.digest,
          registry_url: request.registryUrl
        },
        { deadline },
        (error: grpc.ServiceError | null, response: any) => {
          if (error) {
            reject(this.translateError(error, 'deploy artifact'));
          } else {
            resolve({
              success: response.success,
              message: response.message,
              performerId: response.performer_id
            });
          }
        }
      );
    });
  }

  async removePerformer(performerId: string): Promise<void> {
    this.logger.info(`Removing performer ${performerId}`);
    
    return new Promise((resolve, reject) => {
      const deadline = new Date();
      deadline.setSeconds(deadline.getSeconds() + 60);
      
      this.client.RemovePerformer(
        { performer_id: performerId },
        { deadline },
        (error: grpc.ServiceError | null, response: any) => {
          if (error) {
            reject(this.translateError(error, 'remove performer'));
          } else if (!response.success) {
            reject(new Error(`Failed to remove performer: ${response.message}`));
          } else {
            resolve();
          }
        }
      );
    });
  }

  private translateError(error: grpc.ServiceError, operation: string): Error {
    switch (error.code) {
      case grpc.status.UNAVAILABLE:
        return new Error('Executor service unavailable - check if the executor is running');
      case grpc.status.NOT_FOUND:
        return new Error(`${operation} failed: resource not found`);
      case grpc.status.INVALID_ARGUMENT:
        return new Error(`${operation} failed: ${error.message}`);
      default:
        this.logger.debug(`gRPC error details: ${error.stack}`);
        return new Error(`${operation} failed: ${error.message}`);
    }
  }

  close(): void {
    grpc.closeClient(this.client);
  }
}

export function createExecutorClient(context: Context, logger: Logger): ExecutorClient {
  const address = context.executorAddress || 'executor:9090';
  logger.debug(`Connecting to executor at ${address}`);
  return new ExecutorClient(address, logger);
}

File: hgctl-go/internal/client/executor.go
package client

import (
	"context"
	"fmt"
	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials/insecure"

	"github.com/Layr-Labs/hourglass-monorepo/hgctl-go/internal/logger"
	pb "github.com/Layr-Labs/hourglass-monorepo/ponos/gen/protos/eigenlayer/hourglass/v1/executor"
)

type Client struct {
	conn   *grpc.ClientConn
	client pb.ExecutorServiceClient
	logger logger.Logger
}

func NewExecutorClient(address string, log logger.Logger) (*Client, error) {
	conn, err := grpc.NewClient(address, grpc.WithTransportCredentials(insecure.NewCredentials()))
	if err != nil {
		return nil, fmt.Errorf("failed to create client: %w", err)
	}

	return &Client{
		conn:   conn,
		client: pb.NewExecutorServiceClient(conn),
		logger: log,
	}, nil
}

func (c *Client) DeployArtifact(ctx context.Context, avsAddress, digest, registryName string) (string, error) {
	req := &pb.DeployArtifactRequest{
		AvsAddress:  avsAddress,
		Digest:      digest,
		RegistryUrl: registryName,
	}

	resp, err := c.client.DeployArtifact(ctx, req)
	if err != nil {
		return "", fmt.Errorf("deploy artifact RPC failed: %w", err)
	}

	if !resp.Success {
		return "", fmt.Errorf("deployment failed: %s", resp.Message)
	}

	return resp.DeploymentId, nil
}

func (c *Client) GetPerformers(ctx context.Context, avsAddress string) ([]*pb.Performer, error) {
	req := &pb.ListPerformersRequest{
		AvsAddress: avsAddress,
	}

	resp, err := c.client.ListPerformers(ctx, req)
	if err != nil {
		return nil, fmt.Errorf("list performers RPC failed: %w", err)
	}

	return resp.Performers, nil
}

func (c *Client) RemovePerformer(ctx context.Context, performerID string) error {
	req := &pb.RemovePerformerRequest{
		PerformerId: performerID,
	}

	resp, err := c.client.RemovePerformer(ctx, req)
	if err != nil {
		return fmt.Errorf("remove performer RPC failed: %w", err)
	}

	if !resp.Success {
		return fmt.Errorf("removal failed: %s", resp.Message)
	}

	return nil
}

func (c *Client) Close() {
	if c.conn != nil {
		c.conn.Close()
	}
}
